### ISSUE TITLE
Auth.confirmSignIn throw exception "Invalid device key given." 

### ISSUE THREAD

**Issue by ruiyang** [aws-amplify/amplify-js#2087]
**Describe the bug**

**To Reproduce**
1. Create a cognito pool
2. Create a user with email and phone number verified
3. Enable SMS 2FA
4. Call Auth.signIn. (using the USER_PASSWORD_AUTH).
5. The response will have challengeName="SMS_MFA"
6. Call Auth.confirmSignIn, the promise got rejected with error "Invalid device key given."

There are two calls in confirmSignIn:
1. Confirmed the MFA code - Success
2. Confirm the device - Failed.
The user actually successfully logged in. Because Cognito wants to remember the user 's device, so triggered the second request, which failed. 

**Expected behavior**
Auth.confirmSignIn should run successfully.

**Screenshots**

**Desktop (please complete the following information):**
 - OS: [Mac]

**Smartphone (please complete the following information):**

**Additional context**

***You can turn on the debug mode to provide more info for us by setting window.LOG_LEVEL = 'DEBUG'; in your app.***

Curl copy of the failed request from Chrome:
curl 'https://cognito-idp.ap-southeast-2.amazonaws.com/' -H 'Referer: http://localhost:3000/public/login' -H 'Origin: http://localhost:3000' -H 'X-Amz-Target: AWSCognitoIdentityProviderService.ConfirmDevice' -H 'X-Amz-User-Agent: aws-amplify/0.1.x js' -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.77 Safari/537.36' -H 'Content-Type: application/x-amz-json-1.1' --data-binary '{"DeviceKey":"...","AccessToken":"...","DeviceSecretVerifierConfig":{"Salt":"...","PasswordVerifier":"..."},"DeviceName":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.77 Safari/537.36"}' --compressed

**Comment by jonixj** [comment]
I get the same issue when using phone_number as username. When I use the GUUID as username I do not get this error

**Comment by powerful23** [comment]
@gbenattar how did you get this error? I tried to follows @ruiyang 's steps but couldn't reproduce the error.

**Comment by powerful23** [comment]
@gbenattar when you mean `without 2FA`, you mean to `suppress the second factor during multi-factor authentication (MFA)`?

**Comment by ruiyang** [comment]
@powerful23 Hi, we submitted an aws support ticket for this issue, case 5515598231. The support said:
_The reason behind the ‘InvalidParameterException’ with the message 'Invalid device key given’ while using the ‘USER_PASSWORD_AUTH’ auth flow is that the ‘USER_PASSWORD_AUTH’ does not accept any parameter named device key. Thus, it is not a valid parameter. The device key parameter can only be used with 'USER_SRP_AUTH' and ‘CUSTOM_AUTH’ auth flow. [1]_ 
So based on that I think amplify js should not fire confirm device call if current auth flow is USER_PASSWORD_AUTH.

**Comment by gbenattar** [comment]
Also error messages should be improved.

**Comment by powerful23** [comment]
@ruiyang thanks. I will confirm this with the Cognito team.

**Comment by g-borgulya** [comment]
I have the same issue using CUSTOM_AUTH without MFA. It's a passwordless, SMS-only authentication, where lambda functions take care of sending the SMS and validate the answer. Using amplify's (undocumented) Auth.sendCustomChallengeAnswer for that, which seems to work here as well, as cognito-idp responds with something like:

```
{
  AuthenticationResult: {
    AccessToken: '..',
    ExpiresIn: 3600,
    IdToken: '...',
    NewDeviceMetadata: {
      DeviceGroupKey: '-XXXXX...',
      DeviceKey: 'us-east-2_UUID...'
    },
    RefreshToken: '...',
    TokenType: 'Bearer'
  },
  ChallengeParameters: {}
}
```

But right after this call there's another request to cognito-idp with AccessKey, DeviceKey, DeviceName and DeviceSecretVerifierConfig in the POST body. Which then fails, cognito-idp responds 400 with the message "Invalid device key given.".

I'm not sure what this second request is, it's not straightforward from the amplify / amazon-cognito-identity-js source. I don't know how to prevent that call - or how to correct it if it makes more sense. Nor could I figure out how to make cognito-idp accept/ignore this parameter as an alternative.

Any ideas how to proceed here?

**Comment by remino** [comment]
Just like @g-borgulya, I'm encountering the same issue right now, and that is with Cognito's plain software token MFA, with code copied straight from the examples.

**Comment by DannyHinshaw** [comment]
I'm running into a related issue. We require that the web user uses login with USER_PASSWORD_AUTH _AND_ MFA every time. This includes the initial setup, where we need the user to:

1. Login with temp password
2. Change temp password
3. Retrieve MFA code

Currently I'm getting this error calling the `Auth.completeNewPassword` method. For context, here is the code in question, granted it's still test code at this point, so maybe I'm doing something wrong? If not, I think I've found a ridiculous temp hack around it.

Code producing error:

```typescript
	// Submit login
	const submit = (): any => {
		return Auth.signIn(email, password).then((user) => {
			console.log("THE USER::", user);

			// Have user walk through auth flow
			if (user.challengeName === "NEW_PASSWORD_REQUIRED") {
				console.log("SETTING UP PASSWORD");
				return Auth.completeNewPassword(user, "$H*3hu42o3iHdksj", {})
					.then((loggedUser) => {
                                                // THIS CODE WILL NOT EXECUTE
						console.log("completeNewPassword::", loggedUser);
						return Auth.setupTOTP(loggedUser).then(code => {
							console.log("setupTOTP::code", code);
						})
					}).catch((err) => {
                                                // HERE CATCHES THE INVALID DEVICE KEY ERROR
						console.error("SETUP ERROR:", err);
						return handleError(err);
					});
			}

			// Update admin email to retrieve data after TOTP verify
			return user;
		}).catch(err => {
			setLoading(false);
			return handleError(err);
		});
	};

```

So but the thing is, "completeNewPassword" method actually still accomplishes resetting the password. So technically I could get around this with the following hack **_that should not exist_**:
```typescript
			if (user.challengeName === "NEW_PASSWORD_REQUIRED") {
				console.log("SETTING UP PASSWORD");
				return Auth.completeNewPassword(user, newPass, { email })
					.catch((err) => {
						console.error("SETUP ERROR:", err);
						return Auth.signIn(email, "$H*3hu42o3iHdksj").then((authedUser) => {
							return Auth.setupTOTP(authedUser).then(code => {
								console.log("setupTOTP::code", code);
							});
						});
					});
			}

```

Why...?

**Comment by imranismail** [comment]
Facing this issue too. Any updates on this?

**Comment by dallashuggins** [comment]
We're getting the same issue, but we get the error when calling the `Auth.signIn` method, opposed to `Auth.confirmSignIn`. or `completeNewPassword`. We get the `InvalidParameterException` error "Invalid device key given", but the sign in is successful. When looking at the cognito object we're sending, no device key was being sent. A solution was changing the "Do you want to remember your user's devices?" to "No" - still trying to wrap my head around why, but that stopped the error from occurring.

**Comment by Ch3ck** [comment]
@dallashuggins I had this same issue today and your solution works; I am wondering why too.

**Comment by SahilRanjan05** [comment]
I am using PHP sdk for aws cognito. I am facing problem in suppressing mfa. How to remember device in aws cognito using php sdk. Thanks...

**Comment by gabconcepcioninsurelife** [comment]
> @powerful23 Hi, we submitted an aws support ticket for this issue, case 5515598231. The support said:
> _The reason behind the ‘InvalidParameterException’ with the message 'Invalid device key given’ while using the ‘USER_PASSWORD_AUTH’ auth flow is that the ‘USER_PASSWORD_AUTH’ does not accept any parameter named device key. Thus, it is not a valid parameter. The device key parameter can only be used with 'USER_SRP_AUTH' and ‘CUSTOM_AUTH’ auth flow. [1]_
> So based on that I think amplify js should not fire confirm device call if current auth flow is USER_PASSWORD_AUTH.

base from @ruiyang answer, i switch the auth flow using this guide https://docs.amplify.aws/lib/auth/switch-auth/q/platform/js

**Comment by gabrielmaldi** [comment]
This issue is still valid and should be reopened.

I'm not sure I completely understand this:

> The device key parameter can only be used with 'USER_SRP_AUTH' and ‘CUSTOM_AUTH’ auth flow.

Because I just tested signing in using `USER_PASSWORD_AUTH` which ran my Lambda trigger migration to create the Cognito user, and Amplify sent this request (after successful migration + sign in):

```
x-amz-target: AWSCognitoIdentityProviderService.ConfirmDevice

{
    "DeviceKey": "us-east-1_xxx",
    "AccessToken": "xxx",
    "DeviceSecretVerifierConfig":
    {
        "Salt": "xxx",
        "PasswordVerifier": "xxx"
    },
    "DeviceName": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36"
}
```

With the response being: `200 — { "UserConfirmationNecessary": false }`

*Thus, it seems Cognito is correctly processing a confirm device request for USER_PASSWORD_AUTH (I see the device in the user's details in the User Pool).*

The error occurs when I enable SMS MFA for the user:

1- `InitiateAuth` with the user's credentials returns `{ ChallengeName: "SMS_MFA" }`
2- Amplify calls `RespondToAuthChallenge` with the code the user received via SMS and the response contains `{ NewDeviceMetadata: { DeviceGroupKey: "xxx", DeviceKey: "xxx" } }`
3- Amplify calls `ConfirmDevice` just like it does in the no-mfa-flow I described above, but this time the response is `InvalidParameterException — Invalid device key given.`

So, why is Cognito returning an error this time with the same USER_PASSWORD_AUTH flow, with the only difference being MFA? Is this an issue in Cognito or Amplify? It should be possible to use the migration trigger (which requires USER_PASSWORD_AUTH) and also remember devices (which *does work* except with MFA).

Thanks!

/cc @manueliglesias
